src/persist/src/indexed/mod.rs

```diff
@@ -73,6 +73,60 @@ impl PendingResponse {
     }
 }
 
+struct Pending {
+    // Only drained by drain_pending_writes.
+    pending_writes: Vec<(Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>)>,
+    // Only drained by drain_pending.
+    pending_responses: Vec<PendingResponse>,
+}
+
+impl Pending {
+    fn new() -> Self {
+        Self {
+            pending_writes: Vec::new(),
+            pending_responses: Vec::new(),
+        }
+    }
+
+    fn validate_empty(&self) -> Result<(), Error> {
+        if !self.pending_writes.is_empty() {
+            return Err(Error::from(format!(
+                "still have {} pending writes after draining pending writes, expected 0.",
+                self.pending_writes.len()
+            )));
+        }
+
+        if !self.pending_responses.is_empty() {
+            return Err(Error::from(format!(
+                "still have {} pending responses after draining pending responses, expected 0.",
+                self.pending_responses.len()
+            )));
+        }
+
+        Ok(())
+    }
+
+    fn add_writes(&mut self, updates: Vec<(Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>)>) {
+        self.pending_writes.extend(updates);
+    }
+
+    fn add_response(&mut self, resp: PendingResponse) {
+        self.pending_responses.push(resp);
+    }
+
+    fn writes(&mut self) -> Vec<(Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>)> {
```

Comment:
lets call this take_writes, so it's more obvious what it's doing. ditto for responses